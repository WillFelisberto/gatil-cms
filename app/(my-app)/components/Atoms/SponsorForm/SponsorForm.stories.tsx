import type { Meta, StoryObj } from '@storybook/react';

import { SponsorForm } from './SponsorForm';

const meta: Meta<typeof SponsorForm> = {
  title: 'Atoms/SponsorForm',
  component: SponsorForm,
  tags: ['autodocs']
};

const baseCat = {
  id: 1,
  nome: 'Gato Teste',
  sexo: 'M' as const,
  createdAt: '2025-04-02T21:24:52.909Z',
  updatedAt: '2025-04-03T20:30:41.034Z',
  idade: '1 ano',
  descricao: 'maluca doidinha',
  foto: {
    createdAt: '2025-04-02T20:48:51.116Z',
    updatedAt: '2025-04-02T20:48:51.116Z',
    alt: 'Will',
    filename: 'gato3.png',
    mimeType: 'image/png',
    filesize: 1008532,
    width: 800,
    height: 800,
    focalX: 50,
    focalY: 50,
    sizes: {
      thumbnail: {
        width: 400,
        height: 300,
        mimeType: 'image/png',
        filesize: 324915,
        filename: 'gato3-400x300.png',
        url: '/api/media/file/gato3-400x300.png'
      },

      card: {
        width: 768,
        height: 1024,
        mimeType: 'image/png',
        filesize: 1794522,
        filename: 'gato3-768x1024.png',
        url: '/api/media/file/gato3-768x1024.png'
      },

      tablet: {
        width: null,
        height: null,
        mimeType: null,
        filesize: null,
        filename: null,
        url: null
      }
    },
    id: 67,
    url: '/api/media/file/gato3.png',
    thumbnailURL: '/api/media/file/gato3-400x300.png'
  },
  galeria: [],
  vacinas: [],
  vermifugacoes: [],
  castrado: true,
  dataEntrada: '2025-04-02T12:00:00.000Z',
  adotado: false,
  show: true,
  disponivelParaApadrinhamento: true
};

export default meta;
type Story = StoryObj<typeof SponsorForm>;

export const Default: Story = {
  args: {
    cat: baseCat
  }
};
