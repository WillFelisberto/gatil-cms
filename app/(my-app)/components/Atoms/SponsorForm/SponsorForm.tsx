'use client';

import { Controller, useForm } from 'react-hook-form';
import { IMaskInput } from 'react-imask';
import { toast } from 'react-toastify';

import { Cat } from '@/payload-types';

type SponsorFormData = {
  padrinhoNome: string;
  padrinhoEmail: string;
  whatsapp: string;
  valorMensal: string;
  formaPagamento: string;
  certificadoDigital?: boolean;
  visitas?: boolean;
  aceitaTermos: boolean;
};

type SponsorFormProps = {
  cat: Cat;
  onSuccess(): void;
};

export const SponsorForm = ({ cat, onSuccess }: SponsorFormProps) => {
  const {
    register,
    control,
    handleSubmit,
    formState: { errors, isSubmitting }
  } = useForm<SponsorFormData>();

  const onSubmit = async (data: SponsorFormData) => {
    try {
      const res = await fetch('/api/create-sponsorships', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'x-api-key': process.env.CRON_SECRET || ''
        },
        body: JSON.stringify({
          gato: cat.id,
          gatoNome: cat.nome,
          whatsapp: data.whatsapp,
          valorMensal: data.valorMensal,
          formaPagamento: data.formaPagamento,
          certificadoDigital: data.certificadoDigital,
          visitas: data.visitas,
          aceitaTermos: data.aceitaTermos,
          padrinhoNome: data.padrinhoNome,
          padrinhoEmail: data.padrinhoEmail
        })
      });

      if (!res.ok) throw new Error('Erro ao enviar formulário');
      toast.success('Apadrinhamento enviado com sucesso! 🐾');
      onSuccess();
    } catch {
      toast.error('Erro ao enviar. Tente novamente. 😿');
    }
  };

  return (
    <form
      onSubmit={handleSubmit(onSubmit)}
      className="space-y-6 bg-white w-full mx-auto"
      data-testid="sponsor-form"
      aria-label="Formulário de apadrinhamento"
    >
      {/* Nome */}
      <div>
        <label htmlFor="padrinhoNome" className="block mb-1 font-semibold text-gray-800">
          Seu nome
        </label>
        <input
          id="padrinhoNome"
          type="text"
          placeholder="Ex: João Vítor"
          aria-invalid={!!errors.padrinhoNome}
          aria-describedby={errors.padrinhoNome ? 'padrinhoNome-error' : undefined}
          {...register('padrinhoNome', { required: 'Informe seu nome' })}
          className={`w-full p-2 border rounded-md focus:ring-2 focus:outline-none ${
            errors.padrinhoNome ? 'border-red-500' : 'border-gray-300'
          }`}
          data-testid="input-nome"
        />
        {errors.padrinhoNome && (
          <p id="padrinhoNome-error" className="text-red-500 text-sm mt-1">
            {errors.padrinhoNome.message}
          </p>
        )}
      </div>

      {/* WhatsApp */}
      <div>
        <label htmlFor="whatsapp" className="block mb-1 font-semibold text-gray-800">
          WhatsApp
        </label>
        <Controller
          name="whatsapp"
          control={control}
          rules={{ required: 'Informe seu número de WhatsApp' }}
          render={({ field }) => (
            <IMaskInput
              id="whatsapp"
              {...field}
              mask="(00) 00000-0000"
              unmask={true}
              placeholder="(00) 00000-0000"
              className={`w-full p-2 border rounded-md focus:ring-2 focus:outline-none ${
                errors.whatsapp ? 'border-red-500' : 'border-gray-300'
              }`}
              aria-invalid={!!errors.whatsapp}
              aria-describedby={errors.whatsapp ? 'whatsapp-error' : undefined}
              data-testid="input-whatsapp"
            />
          )}
        />
        {errors.whatsapp && (
          <p id="whatsapp-error" className="text-red-500 text-sm mt-1">
            {errors.whatsapp.message}
          </p>
        )}
      </div>

      {/* Email */}
      <div>
        <label htmlFor="padrinhoEmail" className="block mb-1 font-semibold text-gray-800">
          Seu email
        </label>
        <input
          id="padrinhoEmail"
          type="email"
          placeholder="seuemail@exemplo.com"
          {...register('padrinhoEmail', { required: 'Informe seu email' })}
          className={`w-full p-2 border rounded-md focus:ring-2 focus:outline-none ${
            errors.padrinhoEmail ? 'border-red-500' : 'border-gray-300'
          }`}
          aria-invalid={!!errors.padrinhoEmail}
          aria-describedby={errors.padrinhoEmail ? 'email-error' : undefined}
          data-testid="input-email"
        />
        {errors.padrinhoEmail && (
          <p id="email-error" className="text-red-500 text-sm mt-1">
            {errors.padrinhoEmail.message}
          </p>
        )}
      </div>

      {/* Plano */}
      <div className="flex flex-col sm:flex-row sm:items-start sm:gap-4">
        <label htmlFor="valorMensal" className="font-semibold text-gray-800 min-w-[180px]">
          Plano de contribuição
        </label>
        <div className="flex flex-col flex-1">
          <select
            id="valorMensal"
            {...register('valorMensal', { required: 'Selecione um plano de contribuição' })}
            className={`p-2 border rounded-md focus:ring-2 focus:outline-none ${
              errors.valorMensal ? 'border-red-500' : 'border-gray-300'
            }`}
            data-testid="select-plano"
            aria-invalid={!!errors.valorMensal}
          >
            <option value="">Selecione um plano</option>
            <option value="R$ 29,90 (Básico)">R$ 29,90 (Básico)</option>
            <option value="R$ 59,90 (Premium)">R$ 59,90 (Premium)</option>
            <option value="R$ 99,90 (Master)">R$ 99,90 (Master)</option>
          </select>
          {errors.valorMensal && (
            <p className="text-red-500 text-sm mt-1">{errors.valorMensal.message}</p>
          )}
        </div>
      </div>

      {/* Pagamento */}
      <div className="flex flex-col sm:flex-row sm:items-start sm:gap-4">
        <label htmlFor="formaPagamento" className="font-semibold text-gray-800 min-w-[180px]">
          Forma de pagamento
        </label>
        <div className="flex flex-col flex-1">
          <select
            id="formaPagamento"
            {...register('formaPagamento', { required: 'Selecione a forma de pagamento' })}
            className={`p-2 border rounded-md focus:ring-2 focus:outline-none ${
              errors.formaPagamento ? 'border-red-500' : 'border-gray-300'
            }`}
            data-testid="select-pagamento"
            aria-invalid={!!errors.formaPagamento}
          >
            <option value="">Selecione</option>
            <option value="Pix">Pix</option>
            <option value="Dinheiro">Dinheiro</option>
          </select>
          {errors.formaPagamento && (
            <p className="text-red-500 text-sm mt-1">{errors.formaPagamento.message}</p>
          )}
        </div>
      </div>

      {/* Checkboxes */}
      <div className="space-y-3">
        <label className="flex items-center gap-2 text-gray-700">
          <input
            type="checkbox"
            {...register('certificadoDigital')}
            data-testid="checkbox-certificado"
          />
          Deseja receber certificado digital?
        </label>

        <label className="flex items-center gap-2 text-gray-700">
          <input type="checkbox" {...register('visitas')} data-testid="checkbox-visitas" />
          Gostaria de participar de eventos ou visitas?
        </label>

        <label className="flex items-center gap-2 text-gray-700">
          <input
            type="checkbox"
            {...register('aceitaTermos', { required: true })}
            aria-invalid={!!errors.aceitaTermos}
            data-testid="checkbox-termos"
          />
          <span>
            Li e aceito os{' '}
            <a
              href="/termos-de-apadrinhamento"
              target="_blank"
              rel="noopener noreferrer"
              className="underline text-[#013274] hover:text-[#012050]"
            >
              termos de apadrinhamento
            </a>
          </span>
        </label>
        {errors.aceitaTermos && (
          <p className="text-red-500 text-sm mt-1">É necessário aceitar os termos</p>
        )}
      </div>

      {/* Botão de envio */}
      <button
        type="submit"
        disabled={isSubmitting}
        className="w-full bg-[#013274] text-white font-semibold py-2 rounded-md hover:bg-[#012050] transition"
        data-testid="submit-button"
      >
        {isSubmitting ? 'Enviando...' : 'Apadrinhar agora'}
      </button>
    </form>
  );
};
