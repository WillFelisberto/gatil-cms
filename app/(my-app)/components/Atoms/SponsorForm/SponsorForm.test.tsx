import { toast } from 'react-toastify';

import { Cat } from '@/payload-types';
import { fireEvent, render, screen, waitFor } from '@/tests/test-utils';

import { SponsorForm } from './SponsorForm';

jest.mock('react-toastify', () => ({
  toast: {
    success: jest.fn(),
    error: jest.fn()
  }
}));

const mockCat: Cat = {
  id: '1',
  nome: 'Gato Teste',
  sexo: 'M',
  createdAt: '',
  updatedAt: '',
  idade: '1 ano',
  descricao: 'Gato fofo',
  foto: undefined,
  galeria: [],
  vacinas: [],
  vermifugacoes: [],
  castrado: true,
  dataEntrada: '',
  adotado: false,
  show: true,
  disponivelParaApadrinhamento: true
};

const onSuccessMock = jest.fn();

describe('<SponsorForm />', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should render all required fields', () => {
    const { container } = render(<SponsorForm cat={mockCat} onSuccess={onSuccessMock} />);

    expect(screen.getByTestId('input-nome')).toBeInTheDocument();
    expect(screen.getByTestId('input-email')).toBeInTheDocument();
    expect(screen.getByTestId('input-whatsapp')).toBeInTheDocument();
    expect(screen.getByTestId('select-plano')).toBeInTheDocument();
    expect(screen.getByTestId('select-pagamento')).toBeInTheDocument();
    expect(screen.getByTestId('checkbox-termos')).toBeInTheDocument();
    expect(screen.getByTestId('submit-button')).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  it('shows validation errors when required fields are empty', async () => {
    render(<SponsorForm cat={mockCat} onSuccess={onSuccessMock} />);
    fireEvent.click(screen.getByTestId('submit-button'));

    await waitFor(() => {
      expect(screen.getByText('Informe seu nome')).toBeInTheDocument();
      expect(screen.getByText('Informe seu número de WhatsApp')).toBeInTheDocument();
      expect(screen.getByText('Informe seu email')).toBeInTheDocument();
      expect(screen.getByText('Selecione um plano de contribuição')).toBeInTheDocument();
      expect(screen.getByText('Selecione a forma de pagamento')).toBeInTheDocument();
      expect(screen.getByText('É necessário aceitar os termos')).toBeInTheDocument();
    });
  });

  it('submits the form successfully with correct data', async () => {
    global.fetch = jest.fn(() =>
      Promise.resolve({
        ok: true,
        json: () => Promise.resolve({ message: 'success' })
      })
    ) as jest.Mock;

    render(<SponsorForm cat={mockCat} onSuccess={onSuccessMock} />);

    fireEvent.change(screen.getByTestId('input-nome'), {
      target: { value: 'João Vítor' }
    });
    fireEvent.change(screen.getByTestId('input-email'), {
      target: { value: 'joao@example.com' }
    });
    fireEvent.change(screen.getByTestId('input-whatsapp'), {
      target: { value: '11999999999' }
    });
    fireEvent.change(screen.getByTestId('select-plano'), {
      target: { value: 'R$ 29,90 (Básico)' }
    });
    fireEvent.change(screen.getByTestId('select-pagamento'), {
      target: { value: 'Pix' }
    });
    fireEvent.click(screen.getByTestId('checkbox-termos'));

    fireEvent.click(screen.getByTestId('submit-button'));

    await waitFor(() => {
      expect(onSuccessMock).toHaveBeenCalled();
      expect(toast.success).toHaveBeenCalledWith('Apadrinhamento enviado com sucesso! 🐾');
    });
  });

  it('shows error toast if API fails', async () => {
    global.fetch = jest.fn(() =>
      Promise.resolve({
        ok: false
      })
    ) as jest.Mock;

    render(<SponsorForm cat={mockCat} onSuccess={onSuccessMock} />);

    fireEvent.change(screen.getByTestId('input-nome'), {
      target: { value: 'João Vítor' }
    });
    fireEvent.change(screen.getByTestId('input-email'), {
      target: { value: 'joao@example.com' }
    });
    fireEvent.change(screen.getByTestId('input-whatsapp'), {
      target: { value: '11999999999' }
    });
    fireEvent.change(screen.getByTestId('select-plano'), {
      target: { value: 'R$ 29,90 (Básico)' }
    });
    fireEvent.change(screen.getByTestId('select-pagamento'), {
      target: { value: 'Pix' }
    });
    fireEvent.click(screen.getByTestId('checkbox-termos'));

    fireEvent.click(screen.getByTestId('submit-button'));

    await waitFor(() => {
      expect(onSuccessMock).not.toHaveBeenCalled();
      expect(toast.error).toHaveBeenCalledWith('Erro ao enviar. Tente novamente. 😿');
    });
  });
});
