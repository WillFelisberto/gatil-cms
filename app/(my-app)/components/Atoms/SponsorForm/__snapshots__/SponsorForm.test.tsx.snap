// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SponsorForm /> should render all required fields 1`] = `
<div>
  <form
    aria-label="Formulário de apadrinhamento"
    class="space-y-6 bg-white w-full mx-auto"
    data-testid="sponsor-form"
  >
    <div>
      <label
        class="block mb-1 font-semibold text-gray-800"
        for="padrinhoNome"
      >
        Seu nome
      </label>
      <input
        aria-invalid="false"
        class="w-full p-2 border rounded-md focus:ring-2 focus:outline-none border-gray-300"
        data-testid="input-nome"
        id="padrinhoNome"
        name="padrinhoNome"
        placeholder="Ex: João Vítor"
        type="text"
      />
    </div>
    <div>
      <label
        class="block mb-1 font-semibold text-gray-800"
        for="whatsapp"
      >
        WhatsApp
      </label>
      <input
        aria-invalid="false"
        class="w-full p-2 border rounded-md focus:ring-2 focus:outline-none border-gray-300"
        data-testid="input-whatsapp"
        id="whatsapp"
        name="whatsapp"
        placeholder="(00) 00000-0000"
        value=""
      />
    </div>
    <div>
      <label
        class="block mb-1 font-semibold text-gray-800"
        for="padrinhoEmail"
      >
        Seu email
      </label>
      <input
        aria-invalid="false"
        class="w-full p-2 border rounded-md focus:ring-2 focus:outline-none border-gray-300"
        data-testid="input-email"
        id="padrinhoEmail"
        name="padrinhoEmail"
        placeholder="seuemail@exemplo.com"
        type="email"
      />
    </div>
    <div
      class="flex flex-col sm:flex-row sm:items-start sm:gap-4"
    >
      <label
        class="font-semibold text-gray-800 min-w-[180px]"
        for="valorMensal"
      >
        Plano de contribuição
      </label>
      <div
        class="flex flex-col flex-1"
      >
        <select
          aria-invalid="false"
          class="p-2 border rounded-md focus:ring-2 focus:outline-none border-gray-300"
          data-testid="select-plano"
          id="valorMensal"
          name="valorMensal"
        >
          <option
            value=""
          >
            Selecione um plano
          </option>
          <option
            value="R$ 29,90 (Básico)"
          >
            R$ 29,90 (Básico)
          </option>
          <option
            value="R$ 59,90 (Premium)"
          >
            R$ 59,90 (Premium)
          </option>
          <option
            value="R$ 99,90 (Master)"
          >
            R$ 99,90 (Master)
          </option>
        </select>
      </div>
    </div>
    <div
      class="flex flex-col sm:flex-row sm:items-start sm:gap-4"
    >
      <label
        class="font-semibold text-gray-800 min-w-[180px]"
        for="formaPagamento"
      >
        Forma de pagamento
      </label>
      <div
        class="flex flex-col flex-1"
      >
        <select
          aria-invalid="false"
          class="p-2 border rounded-md focus:ring-2 focus:outline-none border-gray-300"
          data-testid="select-pagamento"
          id="formaPagamento"
          name="formaPagamento"
        >
          <option
            value=""
          >
            Selecione
          </option>
          <option
            value="Pix"
          >
            Pix
          </option>
          <option
            value="Dinheiro"
          >
            Dinheiro
          </option>
        </select>
      </div>
    </div>
    <div
      class="space-y-3"
    >
      <label
        class="flex items-center gap-2 text-gray-700"
      >
        <input
          data-testid="checkbox-certificado"
          name="certificadoDigital"
          type="checkbox"
        />
        Deseja receber certificado digital?
      </label>
      <label
        class="flex items-center gap-2 text-gray-700"
      >
        <input
          data-testid="checkbox-visitas"
          name="visitas"
          type="checkbox"
        />
        Gostaria de participar de eventos ou visitas?
      </label>
      <label
        class="flex items-center gap-2 text-gray-700"
      >
        <input
          aria-invalid="false"
          data-testid="checkbox-termos"
          name="aceitaTermos"
          type="checkbox"
        />
        Li e aceito os termos de apadrinhamento
      </label>
    </div>
    <button
      class="w-full bg-[#013274] text-white font-semibold py-2 rounded-md hover:bg-[#012050] transition"
      data-testid="submit-button"
      type="submit"
    >
      Apadrinhar agora
    </button>
  </form>
</div>
`;
